source/MaterialXCore/Definition.cpp

```diff
@@ -25,13 +25,50 @@ const string Implementation::LANGUAGE_ATTRIBUTE = "language";
 // NodeDef methods
 //
 
-InterfaceElementPtr NodeDef::getImplementation(const string& target) const
+InterfaceElementPtr NodeDef::getImplementation(const string& target, const string& language) const
 {
-    for (InterfaceElementPtr implement : getDocument()->getMatchingImplementations(getName()))
+    InterfaceElementPtr foundElement = nullptr;
+    for (InterfaceElementPtr element : getDocument()->getMatchingImplementations(getName()))
     {
-        if (targetStringsMatch(implement->getTarget(), target))
+        // If language specified then if it's not a node-graph then check both
+        // language and target. Otherwise just check target if the the implementaitons
+        // is a nodegraph.
+        if (!language.empty())
         {
-            return implement;
+            // Check direct implementation
+            ImplementationPtr implementation = element->asA<Implementation>();
+            if (implementation)
+            {
+                // Check for a language match
+                const std::string& lang = implementation->getLanguage();
+                if (lang.length() && lang == language)
+                {
+                    // Check target. Note that it may be empty.
+                    if (targetStringsMatch(implementation->getTarget(), target))
```

Comment:
It seems odd to have multiple code paths that call targetStringsMatch, and I feel this could be simplified by checking the target case at the top of the loop, and then falling down to the language test if the implementation wasn't disqualified by virtue of a target mismatch.  For example:

```
for (InterfaceElementPtr interface : getDocument()->getMatchingImplementations(getName()))
{
    if (!targetStringsMatch(interface ->getTarget(), target))
    {
        continue;
    }

    if (!language.empty())
    {
        ImplementationPtr implement = interface->asA<Implementation>();
        if (implement && implement->getLanguage() != language)
        {
            continue;
        }
    }

    return interface;
}

return InterfaceElementPtr();
```